DM-26297: Change DBB handoff manager to use bbcp

mxk62 opened the pull request.

mxk62 committed:
Expose connection commands

The handoff manager was not allowing one to change the internally used
file transfer tool.  Similarly, only a limited subset of options for
both scp and SSH client used to execute commands remotely could be
changed through configuration file.  To address this issue I exposed
these commands to a user.  Now they can configure what transfer tool
should be used, what options should be set directly in the configuration
file.  If they shoot themselves in the foot in the process, well, they
will have no one to blame but themselves.

mxk62 committed:
Add some more log messages

mxk62 committed:
Change default log level to INFO

mxk62 committed:
Update documentation

MichelleGower commented on doc/lsst.dbb.buffmngrs.handoff/get-started.rst:
```diff
@@ -107,20 +107,64 @@ sites.  For example:
+   the name of the file being transfer and appropriate target location on the
+   endpoint site.
+   
+#. The ``{file}}`` can be replaced with ``{batch}}``.  It will instruct the
```
Are the extra closing braces typos or ...

mxk62 replied:
Yes, it a typo. I'll fix it.

MichelleGower commented on etc/transd.yaml:
```diff
@@ -6,11 +6,13 @@ endpoint:
-  port: 22
+  commands:
+    remote: "ssh -l {user} {host} \"{command}\""
+    transfer: "scp -BCpq {file} {user}@{host}:{dest}"
```
Weren't there discussions about not keeping compression in the command?

MichelleGower commented on python/lsst/dbb/buffmngrs/handoff/manager.py:
```diff
@@ -191,18 +191,20 @@ def main():
     logger.info("Done.")
 
-    logger.info("Starting buffer monitoring...")
+    logger.info("Starting monitoring the buffer...")
```
Wording a little clunky:  Starting to monitor the buffer or Starting the monitoring of the buffer...

MichelleGower commented on python/lsst/dbb/buffmngrs/handoff/manager.py:
```diff
@@ -191,18 +191,20 @@ def main():
         duration = end - start
         logger.info(f"Scan completed: {awaiting.qsize()} file(s) found.")
-        logger.debug(f"Scan completed in {duration:.2f} sec.")
+        logger.debug(f"Scan duration: {duration:.2f} sec.")
```
Couldn't duration be a normal INFO level message?

mxk62 replied:
Sure, it can. I'll "escalate" the log level.

MichelleGower commented on python/lsst/dbb/buffmngrs/handoff/remote.py:
```diff
@@ -67,30 +69,51 @@ class Porter(Command):
+                logger.error(msg)
+                raise ValueError(msg)
+
+        self.batch_mode = False
```
Please add comment here that explains this as well as the replacing batch/file with source.

MichelleGower commented on python/lsst/dbb/buffmngrs/handoff/remote.py:
```diff
@@ -100,54 +123,85 @@ def run(self):
-                # endpoint site.
-                cmd = f"ssh -p {port} {user}@{host} mkdir -p {stage}"
-                status, stdout, stderr = execute(cmd, timeout=self.time)
+            # Group files based on their location.
```
Please add to this comment a very brief statement about why this is being done.

MichelleGower commented on python/lsst/dbb/buffmngrs/handoff/remote.py:
```diff
@@ -100,54 +123,85 @@ def run(self):
+                # Transfer files to the staging area on the endpoint site.
+                relocated = collections.deque()
+
+                batch = [(head, tail, fn) for fn in filenames]
```
Please add comments to this section.  Also confusing because using a variable batch when may not be in batch mode.

MichelleGower commented on python/lsst/dbb/buffmngrs/handoff/remote.py:
```diff
@@ -100,54 +123,85 @@ def run(self):
-                    msg = f"Command '{cmd}' failed with error: '{stderr}'"
-                    logger.warning(msg)
+                tpl = self.cmds["transfer"]
+                for member in batch:
```
Variable names are confusing.  Batch is a "list" of files to be transferred in a single command, but here batch represents a "list" of batches and member is the batch.  Please fix variable names to be consistent and add more comments to this section.

MichelleGower commented on python/lsst/dbb/buffmngrs/handoff/remote.py:
```diff
@@ -100,54 +123,85 @@ def run(self):
+                dest = os.path.join(buffer, tail)
+
+                tpl = self.cmds["remote"]
+                cmd = tpl.format(**self.params, command=f"mkdir -p {dest}")
```
Would make sure to document that the remote command given in the config must work with "mkdir -p" and the later "find" command.

MichelleGower commented on python/lsst/dbb/buffmngrs/handoff/remote.py:
```diff
@@ -100,54 +123,85 @@ def run(self):
-                    cmd = f"ssh -p {port} {user}@{host} mv {source} {target}"
-                    status, stdout, stderr = execute(cmd, timeout=self.time)
+                tpl = self.cmds["remote"]
+                for member in batch:
```
Same variable name consistency and more code comments here.

MichelleGower reviewed (approved):
Just minor code readability things.

mxk62 committed:
Fix clunky wording in a log message

mxk62 committed:
Include timing in selected log messages

mxk62 committed:
Improve code readability of Porter class

Some fragments of the Porter class run() method were clunky and, as a
result, pretty hard to understand.  To make them easier to follow I
added some comments, changed variable names, and made minor refactoring.
